public/app/features/explore/state/query.test.ts

```diff
@@ -265,22 +273,202 @@ describe('reducer', () => {
+            ],
+            queryKeys: ['initialRow-0', 'mockKey-1'],
+          } as unknown as ExploreItemState);
+      });
+    });
+    describe('addQueryRow with mixed datasources disabled', () => {
+      it('should add query row when there is not yet a row', async () => {
+        const exploreId = ExploreId.left;
+        let dispatch: ThunkDispatch, getState: () => StoreState;
+
+        const store: { dispatch: ThunkDispatch; getState: () => StoreState } = configureStore({
+          ...defaultInitialState,
+          explore: {
+            [exploreId]: {
+              ...defaultInitialState.explore[exploreId],
+              queries: [],
+              datasourceInstance: {
+                query: jest.fn(),
+                getRef: jest.fn(),
+                meta: {
+                  id: 'postgres',
+                },
+              },
+            },
+          },
+        } as unknown as Partial<StoreState>);
+
+        dispatch = store.dispatch;
+        getState = store.getState;
+
+        setupQueryResponse(getState());
+
+        await dispatch(addQueryRow(exploreId, 1));
+
+        expect(getState().explore[exploreId].datasourceInstance?.meta?.id).toBe('postgres');
+        expect(getState().explore[exploreId].queries).toHaveLength(1);
+        expect(getState().explore[exploreId].queryKeys).toEqual(['ds1-0']);
+      });
+
+      it('should add another query row if there are two rows already', async () => {
+        const exploreId = ExploreId.left;
+        let dispatch: ThunkDispatch, getState: () => StoreState;
+
+        const store: { dispatch: ThunkDispatch; getState: () => StoreState } = configureStore({
+          ...defaultInitialState,
+          explore: {
+            [exploreId]: {
+              ...defaultInitialState.explore[exploreId],
+              queries: [
+                {
+                  datasource: { type: 'loki', uid: 'ds3' },
+                  refId: 'C',
+                },
+                {
+                  datasource: { type: 'loki', uid: 'ds4' },
+                  refId: 'D',
+                },
+              ],
+              datasourceInstance: {
+                query: jest.fn(),
+                getRef: jest.fn(),
+                meta: {
+                  id: 'loki',
+                },
+              },
+            },
+          },
+        } as unknown as Partial<StoreState>);
+
+        dispatch = store.dispatch;
+        getState = store.getState;
+
+        setupQueryResponse(getState());
+
+        await dispatch(addQueryRow(exploreId, 1));
+
+        expect(getState().explore[exploreId].datasourceInstance?.meta?.id).toBe('loki');
+        expect(getState().explore[exploreId].queries).toHaveLength(3);
+        expect(getState().explore[exploreId].queryKeys).toEqual(['ds3-0', 'ds4-1', 'ds4-2']);
+      });
+    });
+    describe('addQueryRow with mixed datasources enabled', () => {
+      useFeatToggles = {
```

Comment:
I'm not sure if it's mocked properly. This code will have no effect in the tested code as `featureToggles` is mocked as a function that is never executed. So the code that is using the toggle always returns `undefined` (trying to read the property of a function definition / not its results) for `config.featureToggles.exploreMixedDatasource`:

<img width="661" alt="Screenshot 2023-01-24 at 15 24 01" src="https://user-images.githubusercontent.com/745532/214320527-bde3b072-1aec-4fdb-8b9a-39c0beb60f91.png">

Reply:
The test data in single tests and `config.featureToggles.exploreMixedDatasource` weren't mocked properly.
I think I fixed both issues now.

Reply:
To run the code properly before each test it has to be wrapped in `beforeEach()` inside the `describe`. But even if it was wrapped it won't really have any effect because `featureToggles` is mocked as a function (only the entire object is mocked with a factory function, but that applies to the entire object, not the inner props).

For example if you put console.log in `addQueryRow` function:

```typescript
console.log('config.featureToggles.exploreMixedDatasource =', config.featureToggles.exploreMixedDatasource);
if (!config.featureToggles.exploreMixedDatasource || !rootDatasource?.meta.mixed) {
  datasourceOverride = rootDatasource;
}
```

You will see it always produces:

`config.featureToggles.exploreMixedDatasource = undefined`.

You can access the mocked config with regular import and then change it and clean it in before/afterEach:

```typescript
import { config } from '@grafana/runtime';

//...

jest.mock('@grafana/runtime', () => ({
  ...jest.requireActual('@grafana/runtime'),
  getTemplateSrv: () => ({
    updateTimeRange: jest.fn(),
  }),
  getDataSourceSrv: () => {
    return {
      get: (uid?: string) => datasources.find((ds) => ds.uid === uid) || datasources[0],
    };
  },
  config: {
    featureToggles: {
      exploreMixedDatasource: false,
    },
  },
}));

//...

  describe('with mixed datasources enabled', () => {
    beforeEach(() => {
      config.featureToggles.exploreMixedDatasource = true;
    });

    afterEach(() => {
      config.featureToggles.exploreMixedDatasource = false;
    });

//...
```

Reply:
I don't get your point yet. 
`config.featureToggles.exploreMixedDatasource` is set in `beforeEach()` and when running the test in debug mode it works like expected (see below).
<img width="1531" alt="Screenshot 2023-01-30 at 11 41 36" src="https://user-images.githubusercontent.com/48948963/215455823-c9c6e843-d7ef-441a-be83-91dd29e77217.png">
<img width="1532" alt="Screenshot 2023-01-30 at 11 39 25" src="https://user-images.githubusercontent.com/48948963/215455878-b30536fd-2fbe-44f5-87ee-02d4c8186af4.png">

Reply:
I removed __esModule: true and added a afterEach(). in order to complete it and it still works.